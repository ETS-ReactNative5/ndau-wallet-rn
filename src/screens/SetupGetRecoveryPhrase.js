import React, { Component } from 'react'
import {
  View,
  ScrollView,
  Text,
  Linking,
  PixelRatio,
  Platform
} from 'react-native'
import groupIntoRows from '../helpers/groupIntoRows'
import CommonButton from '../components/CommonButton'
import cssStyles from '../css/styles'
import { SafeAreaView } from 'react-navigation'
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen'
import RecoveryDropdown from '../components/RecoveryDropdown'
import Carousel from 'react-native-looped-carousel'
import { Dialog } from 'react-native-simple-dialogs'
import SetupProgressBar from '../components/SetupProgressBar'
import RecoveryPhaseHelper from '../helpers/RecoveryPhaseHelper'
import MultiSafeHelper from '../helpers/MultiSafeHelper'
import UserData from '../model/UserData'
import AppConstants from '../AppConstants'
import SetupStore from '../model/SetupStore'
import FlashNotification from '../components/FlashNotification'
import FontAwesome from 'react-native-vector-icons/FontAwesome'

const DEFAULT_ROW_LENGTH = 3 // 3 items per row
const _ = require('lodash')

class SetupGetRecoveryPhrase extends Component {
  constructor (props) {
    super(props)
    this.NORMAL_MODE_TEXT =
      `To verify your account please verify your twelve-word recovery` +
      `phrase below. Start typing in the box below, then pick the correct suggestion.`
    this.PASSWORD_RESET_MODE_TEXT =
      'To reset your password, please verify your ' +
      'twelve-word recovery phrase. Start typing in the box below, then pick the ' +
      'correct suggestion'
    this.GENESIS_MODE_TEXT =
      `We're almost ready to get you on the ndau blockchain, ` +
      'but we need one last thing from you. \n\nPlease verify your twelve word ' +
      'recovery phrase. Start typing in the box below, then pick the correct suggestion'

    this.state = {
      size: { width: wp('100%'), height: hp('50%') },
      dialogVisible: false,
      recoverPhraseFull: false,
      // recoverPhraseFull: true,
      textColor: '#ffffff',
      confirmationError: false,
      acquisitionError: false,
      stepNumber: 0,
      introductionText: this.NORMAL_MODE_TEXT,
      mode: AppConstants.NORMAL_MODE
    }

    // TODO: you can uncomment the below if you need to do some testing
    // on a known phrase that works in testnet/devnet
    this.recoveryPhrase = ['', '', '', '', '', '', '', '', '', '', '', '']
    // this.recoveryPhrase = [
    //   'coin',
    //   'tortoise',
    //   'approve',
    //   'jar',
    //   'moon',
    //   'reunion',
    //   'utility',
    //   'hello',
    //   'vicious',
    //   'shed',
    //   'property',
    //   'pet'
    // ]
    // this.recoveryPhrase = [
    //   'goat',
    //   'amount',
    //   'liar',
    //   'amount',
    //   'expire',
    //   'adjust',
    //   'cage',
    //   'candy',
    //   'arch',
    //   'gather',
    //   'drum',
    //   'buyer'
    // ]

    this.boxWidth = '30%'
    this.boxHeight = '13%'
    this.rowLength = DEFAULT_ROW_LENGTH
    // if someone has cranked up the font use 1 row instead
    console.log(`PixelRatio.getFontScale is ${PixelRatio.getFontScale()}`)
    if (PixelRatio.getFontScale() > 2) {
      this.rowLength = 1
      this.boxWidth = '100%'
      this.boxHeight = '30%'
      console.log(`boxWidth: ${this.boxWidth} and boxHeight: ${this.boxHeight}`)
    }
  }

  componentWillMount () {
    const mode = this.props.navigation.getParam(
      'mode',
      AppConstants.NORMAL_MODE
    )
    switch (mode) {
      case AppConstants.GENESIS_MODE:
        introductionText = this.GENESIS_MODE_TEXT
        break
      case AppConstants.PASSWORD_RESET_MODE:
        introductionText = this.PASSWORD_RESET_MODE_TEXT
        break
      default:
        introductionText = this.NORMAL_MODE_TEXT
    }

    this.setState({ mode, introductionText })
  }

  addToRecoveryPhrase = (value, index) => {
    this.recoveryPhrase[index] = value
  }

  noRecoveryPhrase = () => {
    this.setState({ dialogVisible: true })
  }

  sendEmail = () => {
    Linking.openURL(
      'mailto:support@oneiro.freshdesk.com?subject=Lost Recovery Phrase'
    )
  }

  _onLayoutDidChange = e => {
    const layout = e.nativeEvent.layout
    this.setState({ size: { width: layout.width, height: layout.height } })
  }

  _generatePage = index => {
    const style = [this.state.size, cssStyles.recoveryPageView]
    if (index === 0) {
      style.push({
        ...Platform.select({
          android: {
            marginLeft: wp('14%')
          }
        })
      })
    }
    return (
      <View style={style} key={index}>
        <Text
          style={[
            cssStyles.wizardText,
            { marginTop: hp('1%'), marginRight: wp('2%') }
          ]}
        >
          {index + 1}.
        </Text>
        <RecoveryDropdown
          addToRecoveryPhrase={this.addToRecoveryPhrase}
          index={index}
          setAcquisitionError={this.setAcquisitionError}
          recoveryPhrase={this.recoveryPhrase}
        />
      </View>
    )
  }

  _generatePages = () => {
    const pages = this.recoveryPhrase.map((phrase, i) => {
      return this._generatePage(i)
    })

    // add one more page to facilitate the carousel functionality
    pages.push(this._generatePage(pages.length))

    return pages
  }

  _checkRecoveryPhrase = async () => {
    return await RecoveryPhaseHelper.checkRecoveryPhrase(
      this.recoveryPhrase.join().replace(/,/g, ' '),
      this.props.navigation.getParam('user', null)
    )
  }

  setAcquisitionError = value => {
    if (value) {
      FlashNotification.showError('Please select a valid word.', false)
    }
    this.setState({ acquisitionError: value })
  }

  confirm = async () => {
    try {
      const { navigation } = this.props

      if (this.state.mode === AppConstants.PASSWORD_RESET_MODE) {
        navigation.navigate('SetupEncryptionPassword', {
          user,
          walletSetupType: navigation.state.params &&
            navigation.state.params.walletSetupType,
          mode: AppConstants.PASSWORD_RESET_MODE,
          recoveryPhraseString: this.recoveryPhrase.join().replace(/,/g, ' ')
        })
        return
      }

      const user = await this._checkRecoveryPhrase()
      if (user) {
        const encryptionPassword = navigation.getParam(
          'encryptionPassword',
          null
        )
        let marketPrice = 0
        // IF we have a password we are fixing up an account from a 1.6 user here
        // so we fixed it up...now save it...and go back to Dashboard
        if (encryptionPassword) {
          try {
            await UserData.loadData(user)
            marketPrice = await OrderNodeAPI.getMarketPrice()
          } catch (error) {
            FlashNotification.showError(error.message)
          }

          await MultiSafeHelper.saveUser(user, encryptionPassword)

          this.props.navigation.navigate('Dashboard', {
            user,
            encryptionPassword,
            walletSetupType: navigation.state.params &&
              navigation.state.params.walletSetupType,
            marketPrice
          })
        } else {
          SetupStore.recoveryPhrase = this.recoveryPhrase
          navigation.navigate('SetupWalletName', {
            user,
            walletSetupType: navigation.state.params &&
              navigation.state.params.walletSetupType
          })
        }
      } else {
        this.setState({
          textColor: '#f05123',
          confirmationError: true
        })
        FlashNotification.showError('Is this the correct recovery phrase? Please correct any errors.')
      }
    } catch (error) {
      console.warn(error)
      this.setState({
        textColor: '#f05123',
        confirmationError: true
      })
      FlashNotification.showError(
        'Is this the correct recovery phrase? Please correct any errors.',
        true
      )
    }
  }

  pushBack = () => {
    this.setState({
      recoverPhraseFull: false,
      confirmationError: false,
      textColor: '#ffffff',
      stepNumber: 0,
    })
  }

  adjustStepNumber = pageIndex => {
    this.setState({ stepNumber: pageIndex })
    if (pageIndex === this.recoveryPhrase.length) {
      this.setState({ recoverPhraseFull: true })
    }
  }

  _renderAcquisition = () => {
    const pages = this._generatePages()

    return (
      <SafeAreaView style={cssStyles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView
            style={cssStyles.contentContainer}
            keyboardShouldPersistTaps='always'
          >
            <SetupProgressBar
              stepNumber={this.state.stepNumber}
              navigation={this.props.navigation}
            />
            <View style={{ marginBottom: 10 }}>
              <Text style={cssStyles.wizardText}>
                {this.state.introductionText}
              </Text>
            </View>
            <View style={{ flex: 1 }} onLayout={this._onLayoutDidChange}>
              <Carousel
                style={this.state.size}
                leftArrowText={'＜'}
                leftArrowStyle={cssStyles.carouselArrows}
                rightArrowText={'＞'}
                rightArrowStyle={cssStyles.carouselArrows}
                // pageInfo
                // pageInfoTextStyle={cssStyles.smallWhiteText}
                // bullets
                arrows
                isLooped={false}
                autoplay={false}
                onAnimateNextPage={this.adjustStepNumber}
                onPageBeingChanged={this.checkIfDone}
              >
                {pages}
              </Carousel>
            </View>
          </ScrollView>
          <View style={cssStyles.footer}>
            <Text
              onPress={this.noRecoveryPhrase}
              style={[cssStyles.linkText, { textAlign: 'center' }]}
            >
              I don't have my recovery phrase
            </Text>
          </View>
        </View>
        <Dialog
          style={{
            fontSize: 18,
            fontFamily: 'TitilliumWeb-Regular'
          }}
          visible={this.state.dialogVisible}
          // title="Missing Recovery Phrase"
          onTouchOutside={() => this.setState({ dialogVisible: false })}
        >
          <View>
            <Text style={cssStyles.blackDialogText}>
              Your recovery phrase is necessary to prove ownership of your ndau. Your wallet cannot
              be restored without it. If you have lost your recovery phrase please contact.
              {' '}
            </Text>
            <Text onPress={this.sendEmail} style={[cssStyles.blueLinkText]}>
              Oneiro concierge support.
            </Text>
          </View>
        </Dialog>
      </SafeAreaView>
    )
  }

  _renderConfirmation = () => {
    const words = groupIntoRows(this.recoveryPhrase, this.rowLength)
    const styles = {
      rowTextView: {
        height: hp(this.boxHeight),
        width: wp(this.boxWidth)
      },
      textStyle: {
        color: this.state.textColor,
        fontSize: 20,
        fontFamily: 'TitilliumWeb-Regular',
        textAlign: 'center'
      }
    }
    let count = 1

    return (
      <SafeAreaView style={cssStyles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView
            style={cssStyles.contentContainer}
            keyboardShouldPersistTaps='always'
          >
            <SetupProgressBar
              stepNumber={this.state.stepNumber}
              navigation={this.props.navigation}
            />
            <View style={{ marginBottom: 10 }}>
              <Text style={cssStyles.wizardText}>
                Is this the correct recovery phrase?{' '}
              </Text>
            </View>
            {words.map((row, rowIndex) => {
              return (
                <View key={rowIndex} style={cssStyles.rowView}>
                  {row.map((item, index) => {
                    return (
                      <View key={index} style={styles.rowTextView}>
                        <Text style={styles.textStyle}>
                          {count++}.{'\n'}
                          {item}
                        </Text>
                      </View>
                    )
                  })}
                </View>
              )
            })}
          </ScrollView>
          <View style={cssStyles.footer}>
            <View style={cssStyles.navButtonWrapper}>
              <CommonButton onPress={() => this.pushBack()} title='Back' />
            </View>
            <View style={cssStyles.navButtonWrapper}>
              <CommonButton onPress={() => this.confirm()} title='Confirm' />
            </View>
          </View>
        </View>
      </SafeAreaView>
    )
  }

  render () {
    console.log(`recoverPhrase is now: ${this.recoveryPhrase}`)

    return !this.state.recoverPhraseFull
      ? this._renderAcquisition()
      : this._renderConfirmation()
  }
}

export default SetupGetRecoveryPhrase
