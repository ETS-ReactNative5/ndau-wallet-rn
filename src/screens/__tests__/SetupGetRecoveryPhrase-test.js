import { mount } from 'enzyme'
import React from 'react'
import { StyleSheet } from 'react-native'
import SetupGetRecoveryPhrase from '../SetupGetRecoveryPhrase'

import renderer from 'react-test-renderer'

describe('testing SetupGetRecoveryPhrase...', () => {
  let styles = StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  })
  const navigation = {
    navigate: jest.fn(),
    getParam: jest.fn(),
    routeName: ''
  }

  it('renders correctly', () => {
    const tree = renderer
      .create(<SetupGetRecoveryPhrase parentStyles={styles} navigation={navigation} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
