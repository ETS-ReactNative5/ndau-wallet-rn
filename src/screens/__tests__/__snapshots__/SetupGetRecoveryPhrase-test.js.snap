// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupGetRecoveryPhrase... renders correctly 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 26.5,
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <RCTScrollView
      keyboardShouldPersistTaps="always"
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginBottom": 93.5,
            }
          }
        />
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 20,
              }
            }
          >
            Is this the correct recovery phrase? 
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              1
              .
              

              goat
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              2
              .
              

              amount
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              3
              .
              

              liar
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              4
              .
              

              amount
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              5
              .
              

              expire
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              6
              .
              

              adjust
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              7
              .
              

              cage
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              8
              .
              

              candy
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              9
              .
              

              arch
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              10
              .
              

              gather
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              11
              .
              

              drum
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 173.5,
                "width": 225,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 20,
                  "textAlign": "center",
                }
              }
            >
              12
              .
              

              buyer
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        style={
          Object {
            "justifyContent": "space-between",
            "marginTop": 26.5,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityTraits="button"
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignContent": "center",
              "backgroundColor": "#4e957a",
              "borderRadius": 3,
              "height": 93.5,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Back
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "justifyContent": "space-between",
            "marginTop": 26.5,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityTraits="button"
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignContent": "center",
              "backgroundColor": "#4e957a",
              "borderRadius": 3,
              "height": 93.5,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Confirm
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
